### PR TITLE
MMF-3443: Open an issue from SonarQube into the IDE

### PR DESCRIPTION
**Pull request by thahnen**
## Summary

With this feature, we want to enable the users to directly jump from an issue they discovered on SonarQube into the IDE to fix it - or to just investigate (already resolved) issues.

## Testing

Testing is done via integration tests for the most common and easy-to-reproduce cases. Manual testing was done for the corner cases not really reproducible by ITs.

### CODE REVIEW

**Review by henryju — APPROVED** [review]
LGTM!

File: its/pom.xml [hunk 1]
```diff
@@ -29,6 +29,15 @@
     <jdk.min.version>11</jdk.min.version>
     <version.jacoco.plugin>0.8.8</version.jacoco.plugin>
   </properties>
+  
+  <dependencies>
+   <dependency>
+      <groupId>com.google.protobuf</groupId>
+      <artifactId>protobuf-java</artifactId>
```
[the remark this answers is not in the document]
Comment (thahnen) [reply its/pom.xml:36]: On `org.sonarlint.eclipse.its.OpenInIdeTest#getFirstIssue(...)` and `org.sonarlint.eclipse.its.OpenInIdeTest#getFirstBranch()`, I use classes from the package `org.sonarqube.ws.*`

Without the import, it failed even to compile because the dependency is required transitively. So the issues, I think, are coming from the orchestrator relying on the _protobuf_ dependency; maybe that can be changed on the orchestrator side?

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/backend/SonarLintEclipseClient.java [hunk 2]
```diff
@@ -334,9 +336,36 @@ public void showSoonUnsupportedMessage(ShowSoonUnsupportedMessageParams params)
...
+    // We were just asked to connect and create a binding, this cannot happen -> Only log the information
+    var bindingOpt = SonarLintCorePlugin.getServersManager().resolveBinding(project);
+    if (bindingOpt.isEmpty()) {
+      SonarLintLogger.get().error("Open in IDE: The project '" + configScopeId
+        + "' removed its binding in the middle of running the action on '" + params + "'");
+      return;
+    }
+    
+    // Handle expensive checks and actual logic in separate job to not block the thread
+    new OpenIssueInEclipseJob(new OpenIssueInEclipseJobParams("Open in IDE", params, project, bindingOpt.get()))
```
[the remark this answers is not in the document]
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/backend/SonarLintEclipseClient.java:367]: Before using `OpenIssueInEclipseJobParams` I had multiple constructors on `OpenIssueInEclipseJob` (some of them were private because only invoked internally).
When I changed it to ask the user to change his preferences (and therefore would invoke the job after pressing **Apply and Close**) I needed something to be easily handed over to the `SonarLintPreferencePage` and I didn't want to save all the information one-by-one but rather as some form of structure.

I see the point, do you have some alternative in mind?

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/job/OpenIssueInEclipseJob.java [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+  private final ISonarLintProject project;
+  private final ResolvedBinding binding;
+  private final boolean recreatedMarkersAlready;
+  private final boolean askedForPreferenceChangeAlready;
+  
+  public OpenIssueInEclipseJob(OpenIssueInEclipseJobParams params) {
+    super(params.name);
+    
+    this.name = params.name;
+    this.params = params.params;
```
[the remark this answers is not in the document]
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/job/OpenIssueInEclipseJob.java:76]: I don't like it either, please see my comment above in general regarding the `OpenIssueInEclipseJobParams` "construct"

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/backend/SonarLintEclipseClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (henryju) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/backend/SonarLintEclipseClient.java:367]: Then maybe rename the class to something like `OpenIssueContext`
